pumpkin-world/src/chunk/anvil.rs

```diff
@@ -141,41 +186,300 @@ impl ChunkReader for AnvilChunkReader {
+                    ),
+                };
+
+                region_file
+                    .seek(SeekFrom::Start(other_offset - at_size as u64))
+                    .unwrap(); // TODO
+                region_file.write_all(&buf).unwrap_or_else(|_| {
+                    panic!(
+                        "Region file r.-{},{}.mca got corrupted, sorry",
+                        region.0, region.1
+                    )
+                });
+                let location_bytes =
+                    &(((other_offset - at_size as u64) / 4096) as u32).to_be_bytes()[1..4];
+                let size_bytes = [(other_size.div_ceil(4096)) as u8];
+                let location_entry = [location_bytes, &size_bytes].concat();
+                location_table[other_table_entry..other_table_entry + 4]
+                    .as_mut()
+                    .copy_from_slice(&location_entry);
+
+                end_index = (other_offset as isize + other_size as isize - at_size as isize) as u64;
+            }
+        } else {
+            for (offset, size) in location_table.chunks(4).filter_map(|v| {
+                let mut offset = BytesMut::new();
+                offset.put_u8(0);
+                offset.extend_from_slice(&v[0..3]);
+                let offset = offset.get_u32() as u64 * 4096;
+                let size = v[3] as usize * 4096;
+                if offset == 0 && size == 0 {
+                    return None;
+                }
+                Some((offset, size))
+            }) {
+                end_index = u64::max(offset + size as u64, end_index);
+            }
+        }
+
+        let location_bytes = &(end_index as u32 / 4096).to_be_bytes()[1..4];
+        let size_bytes = [(bytes.len().div_ceil(4096)) as u8];
+        location_table[table_entry as usize..table_entry as usize + 4]
+            .as_mut()
+            .copy_from_slice(&[location_bytes, &size_bytes].concat());
+
+        // write new location and timestamp table
+
+        region_file.seek(SeekFrom::Start(0)).unwrap(); // TODO
+        if let Err(err) = region_file.write_all(&[location_table, timestamp_table].concat()) {
+            return Err(ChunkWritingError::IoError(err.kind()));
+        }
+        // dbg!(&location_table.iter().map(|v| v.to_string()).join(","));
+
+        region_file.seek(SeekFrom::Start(end_index)).unwrap(); // TODO
+
+        let mut header: BytesMut = BytesMut::with_capacity(5);
+        // total chunk size includes the byte representing the compression
+        // scheme, so +1.
+        header.put_u32(bytes.len() as u32 + 1);
+        // compression scheme
+        header.put_u8(compression as u8);
+        region_file
+            .write_all(&header)
+            .expect("Failed to write header");
+        region_file.write_all(&bytes).unwrap_or_else(|_| {
+            panic!(
+                "Region file r.-{},{}.mca got corrupted, sorry",
+                region.0, region.1
+            )
+        });
+
+        region_file
```

Comment:
This is incorrect, the pad should make the whole file a multiple of 4096, not just add 4096. It should be `len(header) + len(bytes) + len(pad) % 4096 == 0`